form2request/_base.py

```diff
@@ -0,0 +1,220 @@
+        ' not(re:test(., "^(?:submit|image|reset)$", "i"))'
+        " and (../@checked or"
+        '  not(re:test(., "^(?:checkbox|radio)$", "i")))]]',
+        namespaces={"re": "http://exslt.org/regular-expressions"},
+    )
+    values: list[FormdataKVType] = [
+        (k, "" if v is None else v)
+        for k, v in ((e.name, e.value) for e in inputs)
+        if k and k not in keys
+    ]
+    items = data.items() if isinstance(data, dict) else data
+    values.extend((k, v) for k, v in items if v is not None)
+    encoded_values = [
+        (k.encode(), v.encode())
+        for k, vs in values
+        for v in (cast("Iterable[str]", vs) if _is_listlike(vs) else [cast("str", vs)])
+    ]
+    return urlencode(encoded_values, doseq=True)
+
+
+@dataclass
+class Request:
+    """HTTP request data."""
+
+    url: str
+    method: str
+    headers: list[tuple[str, str]]
+    body: bytes
+
+
+def request_from_form(
+    form: FormElement,
+    data: FormdataType = None,
+    /,
+    *,
+    click: None | bool | HtmlElement = None,
+    method: None | str = None,
+) -> Request:
+    """Return a form submission request.
+
+    *form* should be an instance of :class:`lxml.html.FormElement`.
+
+    *data* should be either a dictionary of a list of 2-item tuples indicating
+    the key-value pairs to include in the request as submission data.
+
+    *click* can be any of:
+
+    -   ``None`` (default): the first submission element of the form (e.g. a
+        submit button) is used to build a request for a click-based
+        form submission.
+
+        If no submission elements are found, the request is built for a
+        non-click-based form submission, i.e. a form submission triggered by a
+        non-click event, such as pressing the Enter key while the focus is in
+        a single-line text input field of the form.
+
+    -   ``True`` behaves like ``None``, but raises a :exc:`ValueError`
+        exception if no submission element is found in the form.
+
+    -   ``False`` builds a request for a non-click-based form submission.
+
+    -   A submission element of *form*, to build a request for a form submission
+        based on the clicking of that specific element.
+
+        On forms with multiple submission elements, specifying the right
+        submission element here may be necessary.
+    """
+    try:
+        click_element = _click_element(form, click)
+    except _NoClickables:
+        raise ValueError(
+            f"No clickable elements found in form {form}. Set click=False or "
+            f"point it to the element to be clicked."
+        ) from None
+    if click_element is not None and (enctype := click_element.get("formenctype")):
+        if enctype != "application/x-www-form-urlencoded":
+            raise NotImplementedError(
+                f"{click_element} has formenctype set to {enctype!r}, which "
+                f"form2request does not currently support."
+            )
```

Comment:
What's the browser behavior in this case?

Reply:
Looks like the same as method. An unsupported value gets converted to the default value. So we should raise this only with `multipart/form-data`, and in the case of `text/plain` we could support that, it only changes the `Content-Type` header.

Reply:
Note to self: even if `formenctype` is unsupported but the form defines a supported `enctype`, the default one is used, i.e. `formenctype` is converted to default and retains precedence over `enctype`.